test/image/compare_pixels_test.js

```diff
@@ -80,230 +57,82 @@ function unique(value, index, self) {
+
+    var img0 = PNG.sync.read(fs.readFileSync(base));
+    var img1 = PNG.sync.read(fs.readFileSync(test));
+    var s0, s1, key;
+
+    key = 'width';
+    s0 = img0[key];
+    s1 = img0[key];
+    if(s0 !== s1) {
+        console.error(key + 's do not match: ' + s0 + ' vs ' + s1);
+        fail(mockName);
     }
 
-    function checkImage() {
-        // baseline image must be generated first
-        if(!common.doesFileExist(imagePaths.baseline)) {
-            var err = new Error('baseline image not found');
-            return onEqualityCheck(err, false);
-        }
-
-        /*
-         * N.B. The non-zero tolerance was added in
-         * https://github.com/plotly/plotly.js/pull/243
-         * where some legend mocks started generating different png outputs
-         * on `npm run test-image` and `npm run test-image -- mock.json`.
-         *
-         * Note that the svg outputs for the problematic mocks were the same
-         * and playing around with the batch size and timeout durations
-         * did not seem to affect the results.
-         *
-         * With the above tolerance individual `npm run test-image` and
-         * `npm run test-image -- mock.json` give the same result.
-         *
-         * Further investigation is needed.
-         */
-
-        var gmOpts = {
-            file: imagePaths.diff,
-            highlightColor: 'purple',
-            tolerance: TOLERANCE
-        };
-
-        gm.compare(
-            imagePaths.test,
-            imagePaths.baseline,
-            gmOpts,
-            onEqualityCheck
-        );
+    key = 'height';
+    s0 = img0[key];
+    s1 = img0[key];
+    if(s0 !== s1) {
+        console.error(key + 's do not match: ' + s0 + ' vs ' + s1);
+        fail(mockName);
     }
 
-    function onEqualityCheck(err, isEqual) {
-        if(err) {
-            common.touch(imagePaths.diff);
-            log(err);
-            return cb(false, mockName);
-        }
-        if(isEqual) {
-            fs.unlinkSync(imagePaths.diff);
-            fs.unlinkSync(imagePaths.test);
-        }
+    var width = img0.width;
+    var height = img0.height;
 
-        cb(isEqual, mockName);
-    }
+    var diff = new PNG({
+        width: width,
+        height: height
+    });
 
-    // 525 means a plotly.js error
-    function onResponse(response) {
-        if(+response.statusCode === 525) {
-            log('plotly.js error');
-            return cb(false, mockName);
-        }
-    }
+    var numDiffPixels = pixelmatch(img0.data, img1.data, diff.data, width, height, {threshold: 0.1});
+    if(numDiffPixels) {
+        fs.writeFileSync(imagePaths.diff, PNG.sync.write(diff));
 
-    // this catches connection errors
-    // e.g. when the image server blows up
-    function onError(err) {
-        log(err);
-        return cb(false, mockName);
-    }
+        // mapbox behave differently from run-to-run
+        // skip error on mapbox diff
+        if(mockName.substr(7) === 'mapbox_') continue;
```

Comment:
wait, so we're including mapbox images in these tests, and generating diffs from them, but not allowing those diffs to fail? When do we actually look at these diffs then?

Reply:
We should have a look at `diffs` during the release cycle.

Reply:
I could add a note here after https://github.com/plotly/dev-docs/pull/188 and this PR merged.

Reply:
To filter `mapbox` mocks this should be `substr(0, 7)` not `substr(7)`.

Reply:
Fixes are in b1de0bba71173e732a4717c0dfeec856042d1cf1.
Thanks to @antoinerg comment, now we test everything with zero tolerance except `mapbox_*` and `gl3d_bunny-hull` which are tested with certain tolerance.